chainer/datasets/image_dataset.py

```diff
@@ -149,11 +156,87 @@ def get_example(self, i):
         full_path = os.path.join(self._root, path)
         image = _read_image_as_array(full_path, self._dtype)
 
-        if image.ndim == 2:
-            # image is greyscale
-            image = image[:, :, numpy.newaxis]
         label = numpy.array(int_label, dtype=self._label_dtype)
-        return image.transpose(2, 0, 1), label
+        return _postprocess_image(image), label
+
+
+class MultiZippedImageDataset(dataset_mixin.DatasetMixin):
+    """Dataset of images built from a list of paths to zip files.
+
+    This dataset reads an external image file in given zipfiles. The
+    zipfiles shall contain only image files.
+    This shall be able to replace ImageDataset and works better on NFS
+    and other networked file systems. The user shall find good balance
+    between zipfile size and number of zipfiles (e.g. granularity)
+
+    Args:
+        zipfilenames (list of strings): List of zipped archive filename.
+        dtype: Data type of resulting image arrays.
+    """
+
+    def __init__(self, zipfilenames, dtype=numpy.float32):
+        self._zfs = [ZippedImageDataset(fn, dtype) for fn in zipfilenames]
+        self._zpaths_accumlens = [0]
+        zplen = 0
+        for zf in self._zfs:
+            zplen += len(zf)
+            self._zpaths_accumlens.append(zplen)
+
+    def __len__(self):
+        return self._zpaths_accumlens[-1]
+
+    def get_example(self, i):
+        tgt = bisect.bisect(self._zpaths_accumlens, i) - 1
+
+        lidx = i - self._zpaths_accumlens[tgt]
+        return self._zfs[tgt].get_example(lidx)
+
+
+class ZippedImageDataset(dataset_mixin.DatasetMixin):
+    """Dataset of images built from a zip file.
+
+    This dataset reads an external image file in the given
+    zipfile. The zipfile shall contain only image files.
+    This shall be able to replace ImageDataset and works better on NFS
+    and other networked file systems. If zipfile becomes too large you
+    may consider ``MultiZippedImageDataset`` as a handy alternative.
+
+    Args:
+        zipfilename (str): a string to point zipfile path
+        dtype: Data type of resulting image arrays
+    """
+
+    def __init__(self, zipfilename, dtype=numpy.float32):
+        self._zipfilename = zipfilename
+        self._zf = zipfile.ZipFile(zipfilename)
+        self._zf_pid = os.getpid()
+        self._dtype = dtype
+        paths = filter(lambda x: not x.endswith('/'), self._zf.namelist())
```

Comment:
How about `self._path = [x for x in self._zf.namelist() if not x.endswith('/')]` ?